src/validation.cpp

```diff
@@ -3111,13 +3111,13 @@ bool CChainState::InvalidateBlock(BlockValidationState& state, CBlockIndex* pind
         }
 
         InvalidChainFound(to_mark_failed);
-    }
 
-    // Only notify about a new block tip if the active chain was modified.
-    if (pindex_was_in_chain) {
-        uiInterface.NotifyBlockTip(GetSynchronizationState(IsInitialBlockDownload()), to_mark_failed->pprev);
+        // Only notify about a new block tip if the active chain was modified.
+        if (pindex_was_in_chain) {
+            uiInterface.NotifyBlockTip(GetSynchronizationState(IsInitialBlockDownload()), to_mark_failed->pprev);
```

Comment:
I believe this change will cause the NotifyBlockTip() callbacks to be invoked while holding cs_main the whole time.

Reply:
Thanks! Narrowed the lock to the IBD call: `uiInterface.NotifyBlockTip(GetSynchronizationState(WITH_LOCK(::cs_main, return IsInitialBlockDownload())), to_mark_failed->pprev);`